File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* MergeList(ListNode* l1,ListNode* l2)
    {
        if(l1==NULL)return l2;
        if(l2==NULL)return l1;
        if(l1->val<=l2->val)
        {
            l1->next=MergeList(l1->next,l2);
            return l1;
        }
        else
        {
            l2->next=MergeList(l1,l2->next);
            return l2;
        }
    }
    ListNode* partitionMerge(int start,int end,vector<ListNode*>& lists)
    {
        if(start>end)return NULL;
        if(start==end) return lists[start];

        int mid=start+(end-start)/2;
        ListNode *l1=partitionMerge(start,mid,lists);
        ListNode *l2=partitionMerge(mid+1,end,lists);

        return MergeList(l1,l2);
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int n=lists.size();
        return partitionMerge(0,n-1,lists);
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
/*
    https://leetcode.com/problems/merge-k-sorted-lists/

    Just declared a queue for easy insertion and deletion purpose rather than array
    and ran a loop while length of queue is greater than the 1 everytime I remove two ListNodes
    from queue and add single listNode by merging both of them & finally return the only element of
    the linked list. & and later all code is of the how to merge two linked list
 */
        if(lists.length == 0){
            return null;
        }
        
        Queue<ListNode> queue = new LinkedList<>();
        
        Collections.addAll(queue, lists);
        
        while (queue.size() > 1) {
            
            ListNode first = queue.remove();
            
            ListNode second = queue.remove();
            
            queue.add(merge(first, second));
        }
        return queue.remove();
        
    }
    
    public ListNode merge(ListNode list1, ListNode list2) {

        ListNode dummy = new ListNode(100);

        ListNode temp = dummy;

        ListNode temp1 = list1;
        ListNode temp2 = list2;

        while (temp1 != null && temp2 != null) {

            if (temp1.val <= temp2.val) {

                temp.next = temp1;
                temp1 = temp1.next;

            } else { //

                temp.next = temp2;
                temp2 = temp2.next;

            }

            temp = temp.next;

        }

        if (temp1 == null) {

            temp.next = temp2;

        } else { // temp 2 is null

            temp.next = temp1;

        }

        return dummy.next;

    }
}